R/verify.R

```diff
@@ -0,0 +1,82 @@
+#' Check Code Reproducibility
+#'
+#' Checks whether code in `teal_data` reproduces the stored objects.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' verify(tdata3)
+#' }
+#' @name verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#' @rdname verify
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
```

Comment:
When should verify not be `TRUE`

Let's say the @code slot is modified manually. Is my assumption for the test below correct?

```r
test_that("verify should fail when code is modified manually to corrupt", {
  td1 <- within(teal_data(), a <- 1)
  
  verify(td1)
  
  # Should the verify slot be invalidated once new code is added?
  td2 <- within(teal_data(), e <- 30:40)
  
  # Corrupt code in td2
  td2@code <- td1@code
  
  verify(td2) # Should this lead to NOT verified?
  
  expect_true(td1@verified)
  expect_false(td2@verified)
})
```

Reply:
In your example `td2` will have `@verified` `TRUE` since it was created empty. In an object like that `verify` will automatically return the object without challenging the code.

You are right, this corruption possible. So is setting `td@verified <- TRUE`.

We assume good faith on part of the user because we are unable to prevent bad faith actions.

Reply:
I'm not sure I agree with trusting the slot blindely. User's intuition when reading documentation is that it "Checks whether code in `teal_data` reproduces the stored objects." 

Keeping as is, then the documentation should be changed then to better state this behavior. "Once it's verified it will always be verified".

Adding a parameter to force the check again might be a good thing (in addition to doc update), or detail the hack to force it as it is now.

```r
td2@verified <- FALSE
verify(td2)
```

Reply:
Far be it from me to trust the user blindly :slightly_smiling_face: 

In a normal workflow one would:
1. create data
2. modify data
3. verify data
4. run app
Nowhere would one tinker with slots as that is generally discouraged within S4. This is good faith behavior.

We wanted to have the verification run once because evaluating all the code may take time.

Since code evaluated in the `qenv` is necessarily verified, so once verified, always verified, holds _unless_ one explicitly violates the slots. Therefore, the ability to force verification seems redundant. Adding a note is not a bad idea. Outlining the hack, not necessarily. What would be the purpose?

Reply:
Hey Andre, I think we discussed this already in this issue https://github.com/insightsengineering/teal.data/pull/188#discussion_r1395005794

I also mentioned that in the comment summarizing the PR https://github.com/insightsengineering/teal.data/pull/188#issue-1987535568

> I did not provide a mechanism to block manual change of @verified slot, as we do not have this for @code / @env anyway.


If we ever invent a mechanism preventing `@code` or `@env` to be changed, we should not bother about `@verified` being changed, when those other two are prevented. IMHO if we implement prevention, it should be done for `@code` and `@env`.

Reply:
I did extend the documentation 903870a4d89303404f8832f2e914e3b2f6639a33 - this was a good suggestion that the state is not clear